동기화 메모장 [STEP1] 타코캣(Ldoy)

Ldoy opened the pull request:
안녕하세요 타일러😄 모든 기능이 잘 구현된 상태에서 PR 보내고 싶었는데 너무 늦어질것 같아서 완벽하지 않은 상태지만...PR 보냅니다😭
잘 부탁드립니다🌟🌟🌟
@TaeJoongYoon 

## 코드에 관하여
1.  `CellId`열거형
    - CellId가 한 곳에 있으면 좋을 것 같다는 생각으로 위 열거형을 만들었습니다.  

2. `UIView` Extension에 관하여
    - `UIView`의 위치를 만들어 줄 때 중복적으로 사용하는 Anchor 설정을 `setPosition` 메소드로 만들어 보았습니다. 

3. `CustomDateFormatte` 구조체
    - `dateFormatter` 속성의 경우 `DateFomatter`를 반환하는 클로저 타입으로 만들어 주었습니다.
    - 이유 : `dateFormatter`의 속성을 많이 바꿔주어야 했는데 메서드를 통해서 하기보단 만들어질 때 정해지면 편리할 것 같아서 

4. `MemoDataHolder` 타입
    - 하나의 객체에서 decoding한 데이터 관리할 수 있도록 singleton으로 구현했습니다. 

5. `MainVCTableViewDataSource` 타입
    - `tableView`의 `dataSource`로 `MainVCTableViewDataSource` 인스턴스를 할당해 주었습니다
    - `cellForRowAt`메소드에서 `cell`의 `configure` 메소드에 `CellContentDataHolder` 타입을 통해 `cell`을 구성하는 UI요소들에 반영해야하는 값들을 전달해 주었습니다. 
    - 이유 
        - `cellForRowAt` 메소드 내부에서 `cell` 내부의 content값을 직접 지정하기보단, 값을 가지고 있는 객체를 이용하여(`CellContentDataHolder`) 전달하기만 하는것이 `cell`과 의존성을 낮추는 방향이라고 생각했습니다.
        - Closure, NotificationCenter, delegate 등을 이용해서 데이터를 전달하는 방법들이 있지만 데이터를 가지고 있다가 전달 해주는 단순한 역할이라고 생각해서 CellContentDataHolder 타입을 이용해 보았습니다. 
        - 추후에 cell에 들어가는 data에 대한 분기처리가 필요한 경우 CellContentDataHolder 내부에서 구현하면 좋을 것 같았습니다. 

## 기능구현이 잘 안된 부분과 의문점
1. **tableView의 Delegate를 구분하고 싶었는데 잘 안된 부분**
    - 현재는 PrimaryViewcontroller에서 delegate역할을 하고 있는데 다른 객체가 delegate 역할을 하도록 구현하고자 했습니다. 


2. **cell 스택뷰의 경고창**
    - 스택뷰의 레이아웃을 잘못 설정한 것 같아서 해당 에러를 고치고 싶었는데 고치지 못했습니다. 

| 경고 메세지  | Debugging Consol |
| -------- | -------- | 
![](https://i.imgur.com/OW7LZHk.png)| ![](https://i.imgur.com/TvxFTiu.png)|


<br>3. **다이나믹 타입을 적용했을 때 cell에는 적용이 안되는 부분**
 - Textview에는 잘 적용이 되는데 cell에는 잘 적용이 안됐습니다. 아마 stackView의 문제가 아닐까 생각했는데 타일러의 의견이 궁금합니다.
![](https://i.imgur.com/s8qGglW.gif)

    - 이 문제를 해결하기 위해 아래 코드를 PrimaryViewController에 추가해 보았는데 실패했습니다. 
        ```swift
        tableView.rowHeight = UITableView.automaticDimension
        ```

4. **Cell의 accessoryView에 관하여**
    - 현재는 cell의 accessoryView타입을 지정해 주어서 textView가 있는곳으로 이동하도록 구현했습니다. 
    - 하지만 이렇게 하기 전에 UIButton을 직접 추가해서 구현해보고 싶었고 여러 시도를 해보았는데요! 아래와 같은 상황이 펼쳐졌습니다. 

    | 디버깅콘솔 | 에러 메세지 |
    | -------- | -------- |
    | ![](https://i.imgur.com/P4KghOi.png) |   ![](https://i.imgur.com/0uHGpwR.png) |

    ```swift
    private func makeButtonLayout() {
        self.detailButton = UIButton()
        contentView.addSubview(detailButton)
        detailButton.setAnchor(top: contentView.topAnchor,
                               bottom: contentView.bottomAnchor,
                               leading: nil,
                               trailing: contentView.trailingAnchor)
    }
    ```

    - cell의 높이를지정해주니 UILabel과 stackView는 겹치게 나오진 않았는데 autolayout 경고창은 사라지지가 않더라구요
    - width나 height 값을 주기도하고 leading에 safeAreaLayoutGuide 값을 설정하기도 했는데 결국 안되서 accessory 타입으로 button을 추가해 주었습니다. 
    - 버튼을 추가했는데 왜 stackView와 titleLabel의 vertical이 문제가 생긴건지 알고 싶은데 그러려면 어떤 키워드나 상황을 집중적으로 알아보면 될지 궁금합니다. 
    
2. **StackView의 `leadingAnchor` 에 관하여**
    - 아래처럼 leadingAnchor를 safeAreaLayoutGuide의  leadingAnchor로 변경했는데 잘 잡혔습니다. 그런데 이유를 잘 모르겠습니다!! 제가 어떤 부분을 놓친건지 알고싶습니다


    | 코드 수정 전 | 코드 수정 후 |
    | -------- | -------- |
    | ![](https://i.imgur.com/pJCXMqq.png)|  ![](https://i.imgur.com/zNlmOqg.png)|

     (safeArea 와 view를 구분해 주기 위해 view의 backgroundColor를 빨간색으로 설정했었습니다)
 
    **수정 전 코드**

    ```swift
    private func makeHorizontalStackVeiw() {
            dateAndSubStackView = UIStackView(arrangedSubviews: [self.dateLabel, self.subTitleLabel])
            contentView.addSubview(dateAndSubStackView)
            dateAndSubStackView.translatesAutoresizingMaskIntoConstraints = false
            dateAndSubStackView.axis = .horizontal
            dateAndSubStackView.distribution = .equalCentering

            dateAndSubStackView.spacing = 40
            dateAndSubStackView.trailingAnchor.constraint(equalTo: contentView.trailingAnchor).isActive = true
            dateAndSubStackView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor).isActive = true
            dateAndSubStackView.topAnchor.constraint(equalTo: titleLabel.bottomAnchor).isActive = true

        }
    ```

    **수정 후 코드** : leading 부분만 safeAreaLayoutGuide로 수정 해주었습니다. 
    ```swift
    dateAndSubStackView.leadingAnchor.constraint(equalTo: safeAreaLayoutGuide.leadingAnchor).isActive = true
    ```


6. **`CellId`열거형**
    - CellId가 한 곳에 있으면 좋을 것 같다는 생각으로 위 열거형을 만들었습니다. 그런데 다른 예시코드에선 cell identifier를 cell 타입 내부에서 private 속성으로 가지고 있더라구요. 타일러는 보통 어떻게 하시는지 궁금합니다.

    <br>
### PR을 마치며
아직 autolayout에 관한 개념이 정확하게 잡히지 않은 상태라서 그런지 너무나 당연한 것들인데 질문하는게 많을 것 같다는 생각이 듭니다. 타일러의 너그러운 양해를 구하며.. 잘 부탁드립니다 🙇‍♀️

Ldoy committed:
feat: Set SwiftLint Pakage

Ldoy committed:
feat: Implement JSON Parser Type (struct)

- Add smaple JSON data to Asset

Ldoy committed:
feat: Delete MainStroyBoard

- Add window property in AppDelegate Method
- refactor application method

Ldoy committed:
refactor: Appdelegate 의 application 메소드 수정

- Appdelegate의 application 에 구현했던 것을 SceneDelegate파일에 구현

Ldoy committed:
feat: MemoDataHolder 타입 구현

- sample JSON파일을 디코딩하고 그 데이터를 가지고 있는 타입

Ldoy committed:
refactor: JSONParser, MemoDecodeModel 구조체 리팩토링

- JSONParser의 파라미터 타입 수정 (T -> T.Type)

Ldoy committed:
feat: Layout 구조체 구현

- 테이블뷰의 레이아웃 구성하는 구조체 구현

Ldoy committed:
feat: Cell의 identifier 가지는 CellId 열거형 ㅜㄱ현

Ldoy committed:
feat: MainViewController 에 테이블뷰 추가

Ldoy committed:
feat: UIView에 anchor 만드는 메소드 추가

- mainVC에 해당 메소드로 레이아웃 설정

Ldoy committed:
feat: 테이블뷰 레이블 구현

Ldoy committed:
feat: cell의 stackView의 UILabel 에 hugging 우선순위 설정

- 테이블 뷰 델리게이트 메소드 이용하여 컨텐츠가 겹치는 문제 해결

Ldoy committed:
feat: cell에 accessaryView로 disclosure 타입의 버튼 추가

Ldoy committed:
refactor: Cell style 기능요구서 대로 수정

Ldoy committed:
refactor: MainViewControllerDataSource의 tableView(_:cellForRowAt)메소드 수정

- Cell의 내용물들을 CellContentDataHolder타입 이용해 전달하도록 구현

Ldoy committed:
feat: splitView 연결

Ldoy committed:
feat: Step1 기능 구현 완료(80%)

Ldoy committed:
chore: 여러가지 시도

Ldoy committed:
refactor: regular 사이즈 일때 cell의 수정날짜 업데이트 안 되는 부분 해결\

- textViewDidBeginEditing에서 textViewDidChange 메소드로 변경

Ldoy committed:
refactor: tableView Delegate 메소드 내부 로직 수정

- PrimaryViewController에 giveDataToTextView 메소드 추가

Ldoy committed:
refactor: UIView extension의 pinFullScreen 메소드 삭제

TaeJoongYoon commented on CloudNotes/.swiftlint.yml:
```diff
@@ -0,0 +1,3 @@
+disabled_rules:
+- line_length
+
```
첫 린트 룰이군요 👍

TaeJoongYoon commented on CloudNotes/CloudNotes/Controller/ViewController.swift:
```diff
@@ -6,7 +6,8 @@
 import UIKit
 
-class ViewController: UIViewController {
+@IBDesignable
```
`@IBDesignable`을 추가하신 이유가 있을까요???

Ldoy replied:
`@IBDesignable` 을 추가하면 메인 스토리보드에서 바로 볼 수 있다고 해서 추가했었습니다~!

TaeJoongYoon commented on CloudNotes/CloudNotes/AppLifeCycle/AppDelegate.swift:
```diff
@@ -9,8 +9,18 @@ import CoreData
+        
+        window = UIWindow(frame: UIScreen.main.bounds)
+        
+        let mainVC = MainViewController()
```
변수명은 길어지더라도 끝까지 써주시는게 좋습니다!
`vc`->`viewController`이런식으로요 :)

Ldoy replied:
넵! 유의하겠습니다~😄😄 현재 Scene Delegate에서 이렇게 구현한 부분이 있어서 수정하였습니다!

TaeJoongYoon commented on CloudNotes/CloudNotes/AppLifeCycle/SceneDelegate.swift:
```diff
@@ -12,10 +12,14 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+        window = UIWindow(windowScene: windowScene)
+        let mainVC = MainViewController()
+        window?.rootViewController = mainVC
+        window?.makeKeyAndVisible()
```
이 부분은 이전커밋에 맞춰서 별도의 커밋으로 작성하는게 좋아보이네요 :)

Ldoy replied:
넵! 커밋 나눠서! 변명이지만 혼자하다보니 계속 커밋을 까먹게 되네요.. step2 PR엔 잘 구분해서 할 수 있도록 하겠습니다🤩

TaeJoongYoon commented on CloudNotes/CloudNotes/Data/MemoDataGenerator.swift:
```diff
@@ -0,0 +1,24 @@
+
+class MemoDataHolder {
+    static var list = generateList()
+    private init() { }
```
나중에 싱글톤 객체로 만들기 위한 `private init`일까요???

Ldoy replied:
네! 맞습니다~!

TaeJoongYoon commented on CloudNotes/CloudNotes/Model/JSONParser.swift:
```diff
@@ -19,11 +19,13 @@ enum ParsingError: Error, LocalizedError {
 
 struct JSONParser {
-    static func decoder<T: Decodable> (modelType: T, jsonData: Data) throws -> Result<T, ParsingError> {
+    static func decoder<T: Decodable> (modelType: T.Type, jsonData: Data) throws -> T {
```
이전 커밋에서 함수선언부가 의아했었는데 여기서 의문이 해결되네요 ㅎㅎㅎ
잘 수정된것 같습니다 👍

TaeJoongYoon commented on CloudNotes/CloudNotes/Layout/LayoutGenerator.swift:
```diff
@@ -0,0 +1,19 @@
+        tableView.bottomAnchor.constraint(equalTo: superView.bottomAnchor).isActive = true
+        tableView.rightAnchor.constraint(equalTo: superView.rightAnchor).isActive = true
+    }
+}
```
오호.. 신박한 layout 설계 모델이군요.. 🧐
조금만 더 수정해보면 `UITableView`뿐만 아니라 좀더 제네릭하게 사용할 수 있을것 같아요 ㅎㅎ 
다음 기회에 좀더 수정해보면 좋을것 같네요 ㅎㅎ

Ldoy replied:
지금의 tableView레이아웃은 아래와 같이 수정했습니다~!

```swift
tableView.frame = view.bounds
```

그리고 UIView에서 extension으로 layout을 잡아주는 메소드를 구현했습니다~!

TaeJoongYoon commented on CloudNotes/CloudNotes/PlaceHolder/NamePlace.swift:
```diff
@@ -0,0 +1,19 @@
+
+import Foundation
+
+enum CellId {
```
네이밍 시, 두번째 이후로 오는 축약어는 대문자로 써주는게 좋습니다 :)
`CellId` -> `CellID`

Ldoy replied:
넵! ID인지 Id인지 헷갈렸는데 한 번더 찾아볼걸 그랬네요🥲 수정하여 커밋했습니다~!

TaeJoongYoon commented on CloudNotes/CloudNotes/PlaceHolder/NamePlace.swift:
```diff
@@ -0,0 +1,19 @@
+
+import Foundation
+
+enum CellId {
+    case defaultCell
+    
+    var description: String {
+        switch self {
+        case .defaultCell:
+            return "cell"
+        }
+    }
+}
```
Cell의 identifier를 한곳에서 관리하려는 이유가 있을까요???

Ldoy replied:
PrimaryViewController의 viewDidLoad메소드에서 cell identifier를 이용하는 부분이 있어서 오타의 위험성을 줄이고 싶었습니다~  그리고 나중에 다른 스타일의 cell을 만들수도 있으니 identifier를 모아두면 편리하겠다는 생각이 들었습니다🤓

TaeJoongYoon commented on CloudNotes/CloudNotes/PlaceHolder/NamePlace.swift:
```diff
@@ -0,0 +1,19 @@
+enum CellId {
+    case defaultCell
+    
+    var description: String {
```
`description`을 직접 사용하는것 보단 `identifier`와 같은 변수를 만들어서 사용하는게 직관적일것 같아요 !

Ldoy replied:
아하! 그러네요~! 수정하였습니다~

TaeJoongYoon commented on CloudNotes/CloudNotes/Controller/MainViewController.swift:
```diff
@@ -6,34 +6,16 @@
-        
-        //mainTableView.dataSource = self
+        view.backgroundColor = .white // 배경색
+        tableView.backgroundColor = .systemGray
```
`white`와 `systemGray`는 어떤 속성차이가 있을까요?!

Ldoy replied:
view과 tableView를 구분지어 보고 싶어서 위의 속성을 사용했었습니다! 제가 알고있기론 system이 붙으면 색 구분에 어려움을 가진 분들이 Accesiblity 기능을 통해 색상조절을 할 수 있다고 알고 있습니다!

TaeJoongYoon commented on CloudNotes/CloudNotes/Controller/MainViewController.swift:
```diff
@@ -6,34 +6,16 @@
 
-@IBDesignable
 class MainViewController: UIViewController {
+   let tableView = UITableView()
```
외부에서 접근할 필요가 없다면 `private`을 이용하여 접근제어를 해주세요 :)

TaeJoongYoon commented on CloudNotes/CloudNotes/Controller/MainViewController.swift:
```diff
@@ -6,34 +6,16 @@
 
-@IBDesignable
 class MainViewController: UIViewController {
+   let tableView = UITableView()
```
view를 초기화할때는 lazy하게 초기화하는 방법도 있습니다. 한번 살펴보시면 좋을것 같아요~

Ldoy replied:
넵~! step2 에선 lazy한 초기화 적용할 수 있는지 알아보겠습니다~!

TaeJoongYoon commented on CloudNotes/CloudNotes/Extension/UIView.swift:
```diff
@@ -0,0 +1,29 @@
+import UIKit
+
+extension UIView {
+    func setAnchor(top: NSLayoutYAxisAnchor?, bottom: NSLayoutYAxisAnchor?, left: NSLayoutXAxisAnchor?, right: NSLayoutXAxisAnchor?, layoutMargins: UIEdgeInsets = .zero, size: CGSize = .zero) {
```
`left`, `right`보다 `leading`, 'trailing`을 사용해주세요 :)
왼쪽, 오른쪽이 반대인 나라에서는 헷갈릴수 있겠죠?!

TaeJoongYoon commented on CloudNotes/CloudNotes/Extension/UIView.swift:
```diff
@@ -0,0 +1,29 @@
+        left.flatMap { leftAnchor.constraint(equalTo: $0, constant: layoutMargins.left).isActive = true
+        }
+        
+        right.flatMap { rightAnchor.constraint(equalTo: $0, constant: -layoutMargins.right).isActive = true }
```
들여쓰기 방식을 하나로 통일하면 코드 가독성이 훨씬 좋아질듯 합니다 👍 👍 👍 👍

Ldoy replied:
넵 수정하였습니다~~!

TaeJoongYoon commented on CloudNotes/CloudNotes/Controller/MainViewController.swift:
```diff
@@ -7,15 +7,32 @@
-   let tableView = UITableView()
-
+ let tableView = UITableView()
+   let tableViewDataSource = MainVCTableViewDataSource()
```
`MainVCTableViewDataSource`을 직접만드셨군요! 훌륭한 시도로 보입니다 💯 💯 💯

TaeJoongYoon replied:
들여쓰기를 맞춰주세요 ㅎㅎㅎㅎ

TaeJoongYoon commented on CloudNotes/CloudNotes/Data/MemoDataHolder.swift:
```diff
@@ -11,11 +11,11 @@ class MemoDataHolder {
+            let decodedData = try JSONParser.decoder(modelType: [MemoDecodeModel].self, jsonData: dataAsset)
             return decodedData
         } catch {
             return nil
```
여기서 `return []`을 해주면 `MemoDataHolder.list`을 사용하는 부분에서 불필요한 옵셔널 처리를 안해도 될것같은데 어떻게 생각하세요??

Ldoy replied:
오호! 완전 좋은방법인 것 같습니다~🤩피드백 주신 방향으로 수정해 보았습니다!

TaeJoongYoon commented on CloudNotes/CloudNotes/Extension/UIView.swift:
```diff
@@ -8,7 +8,7 @@
 
 extension UIView {
-    func setAnchor(top: NSLayoutYAxisAnchor?, bottom: NSLayoutYAxisAnchor?, left: NSLayoutXAxisAnchor?, right: NSLayoutXAxisAnchor?, layoutMargins: UIEdgeInsets = .zero, size: CGSize = .zero) {
+    func setAnchor(top: NSLayoutYAxisAnchor?, bottom: NSLayoutYAxisAnchor?, leading: NSLayoutXAxisAnchor?, trailing: NSLayoutXAxisAnchor?, layoutMargins: NSDirectionalEdgeInsets = .zero, size: CGSize = .zero) {
```
여기서 바꿔주셨군요 ㅎㅎㅎ 역시 믿고있었습니다 👍 💯

TaeJoongYoon commented on CloudNotes/CloudNotes/Model/MemoDecodeModel.swift:
```diff
@@ -9,9 +9,7 @@ import Foundation
 struct MemoDecodeModel: Decodable {
-    let item: [Item]
     
-    struct Item: Decodable {
```
이렇게되면 `MemoDecodeModel`을 `Item`으로 바꿔도 되지 않을까요??

Ldoy replied:
네 맞습니다~! 그래서 마지막엔 아래처럼 바꾸어 보았습니다~
```swift
struct MemoDecodeModel: Decodable {
    let title: String
    let body: String
    var lastModified: Int
    
    enum CodingKeys: String, CodingKey {
        case title, body
        case lastModified = "last_modified"
    }
}
```

TaeJoongYoon commented on CloudNotes/CloudNotes/View/MainTableViewCell.swift:
```diff
@@ -0,0 +1,92 @@
+    var dateAndSubStackView: UIStackView!
+    var titleDateSubStackView: UIStackView!
+    var labelsStackView: UIStackView!
+    var wholeStackView: UIStackView!
```
오.. 강제 언래핑은 피해주시는게 좋습니다..! 혹시 이유가 있을까요???

Ldoy replied:
흐극 없습니다ㅜ 옵셔널로 선언하기, 인스턴스 할당해 주기 등의 여러가지 방법 해보다가 수정을 안한 것 같습니다! 수정해서 커밋하였습니다~!

TaeJoongYoon commented on CloudNotes/CloudNotes/View/MainTableViewCell.swift:
```diff
@@ -76,6 +76,9 @@ extension MainTableViewCell {
         dateAndSubStackView = UIStackView(arrangedSubviews: [self.dateLabel, self.subTitleLabel])
+        subTitleLabel.setContentHuggingPriority(.defaultHigh, for: .horizontal)
+        dateLabel.setContentHuggingPriority(.defaultLow, for: .horizontal)
+        dateLabel.setContentCompressionResistancePriority(.defaultHigh, for: .horizontal)
```
`setContentHuggingPriority`과 `setContentCompressionResistancePriority`을 사용하셨군요 ㅎㅎ
오토레이아웃을 좀더 고급스럽게 사용하셨네요! 좋습니다 ㅎㅎㅎ

TaeJoongYoon commented on CloudNotes/CloudNotes/Data/MainVCTableViewDataSource.swift:
```diff
@@ -15,15 +15,15 @@ class MainVCTableViewDataSource: NSObject, UITableViewDataSource {
-        guard let cell = tableView.dequeueReusableCell(withIdentifier: CellId.defaultCell.description, for: indexPath) as? MainTableViewCell else { return UITableViewCell() }
+        
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: CellId.defaultCell.description, for: indexPath) as? MainTableViewCell else { return UITableViewCell()
+        }
```
줄바꿈이 이상하게 된것 같아요!

Ldoy replied:
맞습니다🥲 마음을 급하게 먹었더니 계속 나중에 하자 하고 커밋할 때 수정을 놓친 것 같습니다. 지금은 아래와 같이 구현해 두었습니다~
```swift
guard let cell = tableView.dequeueReusableCell(withIdentifier: CellID.defaultCell.description, for: indexPath) as? MainTableViewCell else {
            return UITableViewCell()
        }
```

TaeJoongYoon commented on CloudNotes/CloudNotes/Data/MainVCTableViewDataSource.swift:
```diff
@@ -15,15 +15,15 @@ class MainVCTableViewDataSource: NSObject, UITableViewDataSource {
-        cell.titleLabel.text = list?.title
-        cell.bodyLabel.text = list?.body
-        cell.accessoryType = .disclosureIndicator
+        let cellContent = CellContentDataHolder(title: list?.title, date: "2020.20.20", body: list?.body)
```
`CellContentDataHolder`와 같이 cell의 configure할수 있는 정보를 구조화시켜서 넘긴게 아주 좋네요 💯 💯 💯 
지금처럼 class나 struct로 넘겨도 되고 Protocol로써 넘겨도 좋은 방식이 될것 같습니다 ㅎㅎ

Ldoy replied:
오호 프로토콜!! 생각해보지 못한 부분인데 고민해 보겠습니다~!

TaeJoongYoon commented on CloudNotes/CloudNotes/View/MainTableViewCell.swift:
```diff
@@ -72,3 +78,17 @@ extension MainTableViewCell {
+    var titleLabelText: String?
+    var dateLabelText: String?
+    var bodyLabelText: String?
+    var accessoryType: UITableViewCell.AccessoryType
```
`UITableViewCell.AccessoryType`이 Data영역에 들어갈지는 좀더 고민이 필요해보여요..!
View의 영역에 들어가보여서 분리해도 괜찮을것 같은데 어떻게 생각하세요???

Ldoy replied:
맞는 말씀입니다 😭😭 현재는 cell의 타입이 바뀔 가능성이 없다고 생각해서 `MainTableViewCell` 에서 초기화 시에 `disclosureIndicator` 로 할당되도록 구현하였습니다~!

TaeJoongYoon commented on CloudNotes/CloudNotes/Controller/PrimaryViewController.swift:
```diff
@@ -9,25 +9,20 @@ import UIKit
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        tableView.reloadData()
+        print(#function)
```
잘못 들어간 라인이 있네요 ㅎㅎ

TaeJoongYoon commented on CloudNotes/CloudNotes/Controller/PrimaryViewController.swift:
```diff
@@ -41,20 +36,34 @@ class PrimaryViewController: UIViewController {
+        let lineBreaker = "\n"
+        let emptyString = ""
+        
+        secondVC.flatMap {
```
단순히 옵셔널을 벗기기 위함이라면, 이럴 때는 `guard`나 `if`문 등으로 옵셔널 언래핑을 하는게 더 나아보여요!

Ldoy replied:
넵! 한 번... 고차함수라는 것을 써보고싶어서 적용해 보았습니다..😂

TaeJoongYoon replied:
ㅎㅎㅎㅎ 좋은 시도네요. 💯

TaeJoongYoon commented on CloudNotes/CloudNotes/Controller/PrimaryViewController.swift:
```diff
@@ -41,20 +36,34 @@ class PrimaryViewController: UIViewController {
+        let emptyString = ""
+        
+        secondVC.flatMap {
+            $0.textView.text = "\(MemoDataHolder.list?[indexPath.row].title ?? emptyString)" + lineBreaker + lineBreaker + "\(MemoDataHolder.list?[indexPath.row].body ?? emptyString)"
```
`text`에 바로 대입보다는 앞서 cell에 값을 넘겨준것처럼 Data를 넘겨주고 `SecondaryViewController`내에서 `text`에 값을 대입하는건 어떨까요??

Ldoy replied:
오호 cell처럼! 좋다고 생각합니다~ 피드백 주신 것 처럼 리팩토링해 보았습니다~!

TaeJoongYoon commented on CloudNotes/CloudNotes/Formatter/DateFormatter.swift:
```diff
@@ -0,0 +1,28 @@
+class CustomDateFormatter {
+    private let dateFormatter = { () -> DateFormatter in
+        let formatter = DateFormatter()
+        formatter.dateStyle = .long
+        formatter.timeStyle = .none
+        formatter.locale = Locale(identifier: "ko_KR")
```
`locale`과 `dateStyle`속성을 잘 이용하셨네요! 훌륭합니다 💯

TaeJoongYoon reviewed (changes requested):
안녕하세요 타코캣! 타일러입니다 :)
완벽하지 않은상태여도 괜찮습니다 ㅎㅎ 코드를 보니 이런저런 시도도 많이 하셨고 고민한 흔적들이 보이네요.
완성하는 것도 중요하지만 이런 시도를 통해서 스스로 많이 배우는게 있다면 그것또한 의미있다고 생각합니다!

---
## 1. Cell identifier
개인적으로는 Cell의 identifier는 Cell에 대한 정보고 외부에서 관리할 의무가 없다고 생각되어서 Cell에서 주체적으로 관리하는걸 선호하는 편인데요. CellId가 한 곳에 있으면 좋을 것 같다는 근거가 좀더 있을까요??? ㅎㅎ

## 2. UIView extension
코멘트에도 적었지만 작업해주신것처럼 특정 view(ex, UITableView)말고 UIView 전체적으로 쓰일수 있는 extension이라면 유용하게 사용할수 있을것 같아요. 하지만 코드의 길이가 길어지고 파라미터가 많아지는 순간 가독성이 떨어지거나 협업시에 어려움이 생길수 있습니다. 이런 부분을 좀더 고민해보시면 좋을것 같아요~ :)

## 3. CustomDateFormatter
`CustomDateFormatter`의 역할이 지금은 단순히 날짜를 포맷팅 하는 역할이라 좀더 많은 기능을 가지면 좋겠네요.
만약 지금 기능을 그대로 유지한다면 별도의 구조체보단 `Date`의 `extension`으로 기능을 만들어도 좋을듯 해요 ㅎㅎ

## 4. MemoDataHolder
`MemoDataHolder`가 현재 구현상으로는 단순히 타입 프로퍼티만 사용하고 싱글톤으로써는 사용되지 않고 있어서 이후 방향성에 따라 수정이 필요해 보여요 ㅎㅎㅎ Data를 관리하는 객체를 설계하신건 아주 좋습니다 💯 이러면 역할분리가 더욱 확실히 되는 장점이 있어요 ㅎㅎ

## 5. MainVCTableViewDataSource
코멘트에도 적어놨지만 별도의 `DataSource`를 만드신 점, Cell에 Data를 넘길때 구조화해서 넘긴 점은 아주 인상깊은 점이었습니다 💯 

## 6. tableView의 Delegate를 구분
지금 `TextViewDelegate`처럼 `UITableViewDelegate`를 상속받는 delegate를 만드시고 행동을 주입받아야 할것 같네요.
그런데 Delegate들을 다 따로 만드시는 이유가 있을까요???
UIKit에서 제공해주는 delegate에 더 필요한 부분이 있거나 override해야하거나 한다면 이점이 있을텐데, 그게 아니라면 delegate를 두 번이상 거쳐야 하기 때문에 오히려 복잡해질수도 있을것 같아서요 ㅎㅎ 타코캣의 의견이 궁금합니다!

## 7. cell 스택뷰의 경고창 + StackView의 leadingAnchor
이건 한가지 이슈로 보이는데요. stackView안에 들어가는 label들의 width가 명확하지 않아서 발생하는 이슈입니다.
두 label의 길이가 가변인데 어느 한쪽의 width를 고정시켜주거나 작업해주신 `contentHugging`이나 `contentCompressionResistance`을 이용해야 합니다. 저희가 가장 우선적이어야 하는건 `dateLabel`의 최소 크기입니다. `dateLabel`의 최소 크기가 정해지면 `bodyLabel`은 가변적으로 width을 알아서 맞추게 되겠죠?! 그래서 아래와 같이 적용해주면 위 이슈들은 잘 해결됩니다 ㅎㅎ
`dateLabel.setContentCompressionResistancePriority(.required, for: .horizontal)`
이 부분이 헷갈리신다면 아래 링크가 도움이 되실것 같아요!

[ios-autolayout](https://academy.realm.io/kr/posts/ios-autolayout/)

## 8. 다이나믹 타입을 적용했을 때 cell에는 적용이 안되는 부분
stackView의 오토레이아웃은 잘 잡아주시고 `tableView.rowHeight = UITableView.automaticDimension`도 잘 설정해주셨는데요!
코드를 보니 cell의 titleLabel topAnchor가 안잡혀있네요 ㅎㅎ topAnchor를 contentView.topAnchor로 해주시면 잘 동작할거에요 :)

## 9. vertical autolayout 문제
일반적으로 vertical관련 문제는 `topAnchor`와 `bottomAnchor`가 잘 설정되어 있는지, 또는 `heightAnchor`가 잘 설정되어 있는지를 확인해 봐야합니다. autolayout은 같은 계층의 뷰간의 연결구조이기 때문에 이러한 문제가 발생하면 해당 뷰만이 아닌 전체적으로 확인해봐야 알 수 있을거에요. 하나라도 어긋나면 전체 구조에 영향을 미치기 때문에 잘 살펴봐야합니다. 그리고 콘솔에 오토레이아웃 워닝이 발생하면 [https://www.wtfautolayout.com](https://www.wtfautolayout.com/)여기서도 도움을 얻을수 있으니 나중에 한번 확인해보세요 :)

---

이 외에도 아직 전체적으로 **강제 언래핑**, **네이밍**, **접근 제어**관련한 사항들이 많이 보이고 있는것 같아요!
시간에 대한 부담없이 차근차근 기본적인 룰을 지키면서 지금까지 배워왔던것들을 잊지 말고 잘 작업해주세요 ㅎㅎㅎ
일주일간 고생하셨습니다 :)

Ldoy committed:
refactor: CellId 열거형의 이름을 CellID로 변경

Ldoy committed:
refactor: MainTableViewCell의 강제언래핑속성 수정

Ldoy committed:
refactor : UIView의 setPosition 메소드 내부 줄바꿈 통일

Ldoy committed:
refactor : SceneDelegate 의 scene 메소드 내부 속성 이름 변경

- splitVC 에서 splitViewController로

Ldoy committed:
refactor: CellID열거형 속성 description을 identifier로 변경

Ldoy committed:
refactor: MemoDataHolder에서 MemoData로 이름 수정

Ldoy committed:
feat: MemoDataHolder 클래스 추가

- Secondary ViewController로 메모 정보를 전달하는 역할을 하는 MemoDataHolder 구현

Ldoy committed:
refactor: secondaryViewController로 indexPath, tableView전달

- TableViewIdexPathHolder 이용

Ldoy committed:
refactor: PrimaryViewController의 giveDataToSecondaryVC 메소드 리팩토링중

Ldoy committed:
reafactor : PrimaryViewController에서 SecondaryViewController로 데이터 전달하는 로직 수정

- PrimaryViewController 의 giveDataToSecondaryVC 메소드 수정
- SecondaryViewController에 delegate 채택
- TableViewIdexPathHolder에 textView의 text 속성 초기화할 속성 추가
- TextViewDelegate 파일 삭제

Ldoy committed:
refactor: SecondaryViewController의  configure 메소드 수정하여 '마지막수정날짜 업데이트 기능' 완료

- SecondaryViewController의 self 속성 명시
- SceneDelegate 내부의 줄바꿈 수정

Ldoy committed:
refactor : SecondaryViewController의 configure메소드 위치 변경

- SecondaryViewController 내부 속성에 self 키워드 추가
- updateTextViewText 메소드 추가

Ldoy committed:
refactor: CustomDateFomatter의 내용을  DateFormatter의 extension으로 구현

- Extension 그룹에 DateFormatter 파일 추가
- CustomDateFormatter 삭제
- MainVCTableViewDataSource에 CustomDateFomatter로 구현한 내용 수정

Ldoy committed:
refactor: MemoData에서 singleton패턴이 아니라 타입프로퍼티가 값을 가지고 있도록 함

- MemoData의 private inti삭제
- reference 삭제해서 남아있던 CellId파일 삭제 진행

Ldoy committed:
refactor : CellContentDataHolder의 역할에 맞지않은 accessorytype속성 제거

- MainTableViewCell의 init 메소드에서 accessoryType 할당함

Ldoy commented:
>1. Cell identifier
개인적으로는 Cell의 identifier는 Cell에 대한 정보고 외부에서 관리할 의무가 없다고 생각되어서 Cell에서 주체적으로 관리하는걸 선호하는 편인데요. CellId가 한 곳에 있으면 좋을 것 같다는 근거가 좀더 있을까요??? ㅎㅎ

아하 Cell에관한 정보이니 Cell이 관리하는게 더 맞는 방향일 수도 있겠네요! 처음 저렇게 구현했던 이유는 `PrimaryViewController`의 `viewDidLoad`메소드에서 `cell identifier`를 이용하는 부분이 있어서 오타의 위험성을 줄이고 싶었습니다~ 그리고 나중에 다른 스타일의 cell을 만들수도 있으니 `identifier`를 모아두면 편리하겠다는 생각이 들었습니다. 그런데 답변을 작성하면서 생각해보니 '굳이 CellId를 구현했야 했나..??'' 라는 생각이 드네요😂 조금 더 고민해서 두 번째 PR 전에 수정해 보도록 하겠습니다

> 2. UIView extension
코멘트에도 적었지만 작업해주신것처럼 특정 view(ex, UITableView)말고 UIView 전체적으로 쓰일수 있는 extension이라면 유용하게 사용할수 있을것 같아요. 하지만 코드의 길이가 길어지고 파라미터가 많아지는 순간 가독성이 떨어지거나 협업시에 어려움이 생길수 있습니다. 이런 부분을 좀더 고민해보시면 좋을것 같아요~ :)

넵 그래서 `setPosition`이라는 메소드로 변경하고 내부에서 top, bottom, trailing, leading을 정할 수 있도록 변경해 보았습니다.(제가 커밋에 명시하지 않아서 아마 발견하기 어려우셨을 것 같습니다🥲) 

>3. CustomDateFormatter
CustomDateFormatter의 역할이 지금은 단순히 날짜를 포맷팅 하는 역할이라 좀더 많은 기능을 가지면 좋겠네요.
만약 지금 기능을 그대로 유지한다면 별도의 구조체보단 Date의 extension으로 기능을 만들어도 좋을듯 해요 ㅎㅎ

아하 `extension`으로 구현하는 부분을 생각해보지 못했네요! 타일러 말씀처럼 추가적인 기능이 있지 않은 이상 타입을 따로 가질 필요는 없을 것 같습니다~!해당 부분 리팩토링해서 커밋해 두었습니다~

>4. MemoDataHolder
MemoDataHolder가 현재 구현상으로는 단순히 타입 프로퍼티만 사용하고 싱글톤으로써는 사용되지 않고 있어서 이후 방향성에 따라 수정이 필요해 보여요 ㅎㅎㅎ Data를 관리하는 객체를 설계하신건 아주 좋습니다 💯 이러면 역할분리가 더욱 확실히 되는 장점이 있어요 ㅎㅎ

제가 싱글톤의 개념을 혼동했었던 것 같습니다!singleTon은 하나의 인스턴스를 만드는 것을 보장하는 패턴이고 MemoDataHolder를 만든 이유는 한 곳에서 데이터를 관리하고자 했던 부분인데 잘못된 방향으로 구현했다는 생각이듭니다~! `private init() { }` 부분을 삭제하였습니다!

>5. MainVCTableViewDataSource
코멘트에도 적어놨지만 별도의 DataSource를 만드신 점, Cell에 Data를 넘길때 구조화해서 넘긴 점은 아주 인상깊은 점이었습니다 💯


으핫 감사합니다🤩


>6. tableView의 Delegate를 구분
지금 TextViewDelegate처럼 UITableViewDelegate를 상속받는 delegate를 만드시고 행동을 주입받아야 할것 같네요.
그런데 Delegate들을 다 따로 만드시는 이유가 있을까요???
UIKit에서 제공해주는 delegate에 더 필요한 부분이 있거나 override해야하거나 한다면 이점이 있을텐데, 그게 아니라면 delegate를 두 번이상 거쳐야 하기 때문에 오히려 복잡해질수도 있을것 같아서요 ㅎㅎ 타코캣의 의견이 궁금합니다!

처음에 이렇게 구현하고자 했던 시작은 'MVC 패턴에서 ViewController의 역할이 무엇인가' 라는 질문이었습니다. 

여러 캠퍼와 이야기도 나누고 여러 글을 보며 ViewController에서는 비지니스 로직이 구현되어야 한다는 생각이 들었습니다.
(제가 이해한 비지니스 로직은 'Model과 View를 이용하여 내부에 구현한 규칙'입니다.)

이런 생각 끝에 ViewController가 View의 Delegate 역할을 하는 것이 맞을까? 하는 의문이 들었습니다. 

그래서 최대한 delegate, dataSource에 대한 타입을 따로만들고 해당 타입의 인스턴스를 할당해주어야한다는 생각이 강했던 것 같습니다. 하지만 지금처럼 `UIKit에서 제공해주는 delegate에 더 필요한 부분이 있거나 override해야하거나 한다면 이점` 이 있는게 아니라면 굳이 두 번 거칠 필요없이 해당 컨트롤러에 delegate역할을 부여하는 것이 오히려 더 생산적인 방향이 아닐까 하는 생각이 들었습니다. delegate, datasource등을 분리하는것을 모든 상황에 적용해야하는 법칙처럼 생각했던 것 같아요 😂

Ldoy committed:
refactor: CellContentDataHolder에서 사용하지 않는 dateText속성 제거

- 의존관계 명시적으로 볼 수 있도록 self 키워드 추가
(Controller, Cell, Holder 그룹의 파일)

Ldoy committed:
refactor: DataHolder들의 경우 참조타입이 아니어도 되기 때문에 strcut으로 변경

- DataHolder그룹 내부 파일의 모든 타입을 struct 으로 변경

Ldoy committed:
-chore: 사용하지 않는 DateFormatter 그룹 삭제

Ldoy committed:
refactor: MemoData의 return값 옶셔널 타입 반환하지 않도록 메서드 내부 로직 수정

-  nil 대신 [] 를 반환하도록 함
- MemoData를 가지고 있던 곳에서 옵셔널 처리 제거함

TaeJoongYoon commented on CloudNotes/CloudNotes/Data/MemoDataHolder.swift:
```diff
@@ -6,3 +6,15 @@
+class MemoDataHolder {
+    var testViewText: String?
+    var indexPath: IndexPath?
+    var tableView: UITableView?
```
이 두가지 프로퍼티를 보면 `tableView`에 굉장히 의존적인 걸로 보이는데요.
옵셔널로 받지 않아도 괜찮아 보입니다! 옵셔널로 설정한 이유가 있으실까요??

Ldoy replied:
초기엔 addButton을 눌렀을 때 indexPath와 tableView를 전달하지 않도록 하기위해 옵셔널로 처리했었습니다~! 지금은 옵셔널 처리가 필요 없을 것 같아서 수정하였습니다!!

TaeJoongYoon commented on CloudNotes/CloudNotes/View/Controller/PrimaryViewController.swift:
```diff
@@ -41,13 +41,8 @@ extension PrimaryViewController {
-//            $0.textViewDelegate.tableView = tableViewIndexPathHolder.tableView
-        }
+        secondVC?.textViewDelegate.holder = tableViewIndexPathHolder
+        secondVC?.textView.text = "\(MemoData.list?[indexPath.row].title ?? emptyString)" + lineBreaker + lineBreaker + "\(MemoData.list?[indexPath.row].body ?? emptyString)"
```
이 `text`내용까지 `TableViewIdexPathHolder`와 같은 폼에 담아서 넘겨주고
`secondVC`에서 값을 설정하는게 좋아보여요 ㅎㅎㅎㅎ

TaeJoongYoon commented on CloudNotes/CloudNotes/View/Controller/SecondaryViewController.swift:
```diff
@@ -9,17 +9,21 @@ import UIKit
+    
+    func configure(_ holder: TableViewIdexPathHolder) {
+        textView.text = holder.textViewText
+    }
```
아주 좋습니다 💯 💯

TaeJoongYoon commented on CloudNotes/CloudNotes/Data/Holder/CellContentDataHolder.swift:
```diff
@@ -7,7 +7,7 @@
 import UIKit
 
-class CellContentDataHolder {
+struct CellContentDataHolder {
```
좋은 고민의 결과네요 👍 💯

TaeJoongYoon reviewed (approved):
안녕하세요 타코캣, 타일러입니다~~
STEP1 잘 마무리 해주셨네요 ㅎㅎㅎ 많은 질문 주시고 이런저런 시도도 많이 하신것 같아서 좋았던것 같습니다 ㅎ
많은 질문을 주셨지만 그에맞는 대답을 해드렸는지 생각하게 되네요 ㅎㅎ
STEP2도 잘부탁드리겠습니다 :)

---
##  ViewController가 View의 Delegate 역할

타코캣의 의견에도 공감합니다. 정말 세분화해서 역할을 분리해야한다면 그렇게 하는것도 맞는 방법일거에요. 
이번 프로젝트에서 그런 부분을 더 연습하고 싶으시다면 `Delegate`뿐만 아니라 `View`와 `Controller`까지 분리해서 개발해보시면 더욱 좋을것 같습니다 ㅎㅎㅎ

하지만, 개발하다보시면 느끼시겠지만 이렇게 하는게 이론적으로는 맞을수 있지만 협업과 유지보수에 용이한 지에 대한 고민이 생기실 겁니다. 그런 고민이 생기면 어떻게 해결해야 할 지, 어떤 방법으로 개선해나가는게 좋을 지도 한번 생각해보시면 더욱 많은 도움이 되실것 같아요 ㅎㅎㅎ

적어도 저는 코드에는 정답이 없다고 생각합니다. 지금 프로젝트를 하시면서 많은 시도와 고민을 해보세요! 그러면 타코캣만의 기준과 노하우가 생길거라 믿습니다 ㅎㅎ

---
